### ISSUE TITLE
Thoughts on miles/kwh?

### ISSUE THREAD

**Issue by disruptivepatternmaterial** [bretterer/home-assistant-rivian#155]
As I as I am sure everyone knows the trucks report miles/kwh I the '15 min" view and in the details of the operation of the vehicle. But I cannot figure out how to calculate this from the data this integration returns. Has anyone been able to do that? Thought I would ask before I dig through the code / lib under it to see if this is possible or available...

**Comment by jrgutier** [comment]
Not directly. You can do some math between batteryCapacity, SOC, and the
Odometer sensors to get it.

--

J.R. Gutierrez

**Comment by AlanJohnWilliams** [comment]
I started playing around with this using Influxdb (FLUX) and Grafana. I've got two dashboards - a status focused dashboard (https://gist.github.com/AlanJohnWilliams/ca789ed2b2313c82369101eb6a36ecc7) and a stats focused dashboard (https://gist.github.com/AlanJohnWilliams/9e6cce34412d3f4eb6b21df197c229c2). Feel free to lift out the FLUX queries I have for mi/kWh. 

Some notes on this:

1. Edit the vehicleName dashboard for the name (entity prefix) for your vehicle. The dashboards should support multiple Rivians in the family. 
2. I have hard coded 130 kWh battery capacity in some spots. I have older Rivian integration data that exists prior to this being reported by the integration. You may need to fix that hard coded if you have standard or max pack. 
3. As for efficiency reporting, the resolution of time points is too sparse to get what you get on the dash. But still useful for trip level perspectives. 
4. Not tested extensively, so let me know if you see anything off. 

Stats Dashboard:
<img width="2245" alt="image" src="https://github.com/bretterer/home-assistant-rivian/assets/35285037/11222b2f-4ae1-4b9b-a277-ff915f566f08">

Status Dashboard:
<img width="2244" alt="image" src="https://github.com/bretterer/home-assistant-rivian/assets/35285037/bc2c1236-451f-4ad3-93d4-255b42245be0">

**Comment by AlanJohnWilliams** [comment]
RE the dashboards, my influxdb and Grafana instances are running outside of Home Assistant (not as plugins). The main thing is you may need to jump through some hoops to enable FLUX on the influxdb plugin.

**Comment by disruptivepatternmaterial** [comment]
> RE the dashboards, my influxdb and Grafana instances are running outside of Home Assistant (not as plugins). The main thing is you may need to jump through some hoops to enable FLUX on the influxdb plugin.

Yeah mine too because everything else in my house dumps into influx (which I cannot stand and am working to move toward a normal database so I can do normal rdbms stuff, like modify data, rename thing, fix mistakes, etc...).

So those dashboards are kind of cool, I will give them a shot. I was hoping someone would have solved the math, because what you need is to look for + and - derivative to detect charging and discharging, then take the max value of the charge in between those intervals. If you have those values you can average them divide the miles by that number and boom you have the same data as the dash. I am a tad perplexed why or how the truck has this data point and clearly knows the lifetime charge but doesnt report it out. Seems vastly more useful and obvious than 80% of the data the API is reporting.

**Comment by AlanJohnWilliams** [comment]
Yeah, it would be nice if they reported it out directly! I'm still not 100% confident in the underlying data and assumptions made to get the mi/kWh results. In flux I'm able to get diff on the state of charge for each interval and then multiply it by the battery capacity to get the kWh used. I also sum the kW added from charging over the same interval to then calculate the mi/kWh. I calculate mi/kWh overall (includes phantom drain) and also calculate the value specifically associated with vehicle motion which required a bit more work to filter based on power_state and then split the SoC diff between "Go" state and all other states.

**Comment by tmack8001** [comment]
> why or how the truck has this data point and clearly knows the lifetime charge but doesnt report it out.

Lifetime  Efficiency is reported by the truck. We used to have that before an API deprecation and added API authorization removed it from our view since there currently isn't no need for that data within the Android/iOS mobile clients - we only have visibility to the data that these apps have permission to request. The trucks report much more than what we can see.

**Comment by p3ck** [comment]
I would be curious to see how others have implemented this.  This is what I have in HA:

I have a template sensor that calulates kw
{{ states('sensor.jade_battery_state_of_charge')| float / 100 * states('sensor.jade_battery_capacity') | float }}

I then crated a template trigger which feeds off that:

`
- trigger:
    trigger: state
    entity_id: sensor.jade_power
  condition:
    condition: template
    value_template: >
       {% set power_test = trigger.from_state.state | float(0) - trigger.to_state.state | float(0) %}
       {{ power_test > 0 or is_state('binary_sensor.jade_charging_status', 'off') }}
  sensor:
    - name: Jade Power Test
      unique_id: '1735999774'
      unit_of_measurement: kWh
      device_class: energy
      state: >
        {% set power_test = trigger.from_state.state | float(0) - trigger.to_state.state | float(0) %}
        {{ states('sensor.jade_power_test') | float(0) + power_test }}
      state_class: total
`

Called jade_power_test since I'm still experimenting....  This gives me kwh over time...  and I bring that into grafana to chart it out along with miles driven...  This way I can see many kwh I use on a particular trip.  I simply zoom in or our of the data to show the difference in kwh.

![Image](https://github.com/user-attachments/assets/0bfe6fe6-7ff5-47de-80ab-d71f3510b459)

![Image](https://github.com/user-attachments/assets/c43d51a3-8ef9-42a6-b73f-7533e414a392)
